src/mesh/p4est_mesh.jl

```diff
@@ -0,0 +1,273 @@
+"""
+    P4estMesh{NDIMS, RealT<:Real} <: AbstractMesh{NDIMS}
+
+An unstructured curved mesh based on trees, which uses the C library p4est
+to manage trees and mesh refinement.
+
+!!! warning "Experimental code"
+    This mesh type is experimental and can change any time.
+"""
+mutable struct P4estMesh{NDIMS, RealT<:Real, NDIMSP2} <: AbstractMesh{NDIMS}
+  p4est                 ::Ptr{p4est_t}
+  p4est_mesh            ::Ptr{p4est_mesh_t}
```

Comment:
I'd prefer to only have one mesh. Would something like `Union{Ptr{p4est_t}, Ptr{p8est_t}}` be type stable?

Reply:
No, I don't think so. Also, in many places you'd have to call different functions depending on whether you are in 2d or 3d, right? So you'd have a lot of "if 2d then do this, otherwise that", which is something we have managed to avoid in Trixi so far.

Reply:
It might be possible to use only one parametric type if you add the type of the pointer as type variable. However, as @sloede pointed out, it would be good to keep a good overview and no inner loops with if blocks checking the dimensions. If you find yourself in a situation where the functions you need to call have different names but basically the same signature, it would be good to add such a function as higher-level function in P4est.jl, using dispatch to distinguish the underlying C function names.